rosdoc2/verbs/scan/impl.py

```diff
@@ -0,0 +1,217 @@
+import multiprocessing as mp
+import os
+import signal
+import sys
+import threading
+import time
+
+from catkin_pkg.packages import find_packages_allowing_duplicates
+from rosdoc2.verbs.build.impl import main_impl as build_main_impl
+from rosdoc2.verbs.build.impl import prepare_arguments as build_prepare_arguments
+
+logging.basicConfig(
+    format='[%(name)s] [%(levelname)s] %(message)s', level=logging.INFO)
+logger_scan = logging.getLogger('rosdoc2.scan')
+
+goptions = None
+# Setting the BATCH_SIZE or MAX_PACKAGES to a smaller value may be useful in debugging
+# this module, to reduce run time or isolate sections that cause hangs.
+BATCH_SIZE = 10000
+MAX_PACKAGES = 10000
+WATCHDOG_TIMEOUT = 15 * 60  # Seconds
+
+
+def main(options):
+    """Execute the program, catching errors."""
+    try:
+        return main_impl(options)
+    except Exception as e:  # noqa: B902
+        if options.debug:
+            raise
+        else:
+            sys.exit(str(e))
+
+
+class Struct:
+    """Wrap argparse options to allow copies."""
+
+    def __init__(self, **entries):
+        """Create a dictionary from option entries."""
+        self.__dict__.update(entries)
+
+
+def prepare_arguments(parser):
+    """Add command-line arguments to the argparse object."""
+    # Wrap the builder arguments to include their choices.
+    build_prepare_arguments(parser)
+
+    # Additional options for scan
+    parser.add_argument(
+        '--timeout',
+        '-t',
+        default=WATCHDOG_TIMEOUT,
+        help='maximum time in seconds allowed per package',
+    )
+    parser.add_argument(
+        '--max-packages',
+        '-m',
+        default=MAX_PACKAGES,
+        help='maximum number of packages to process'
+    )
+    return parser
+
+
+def main_impl(options):
+    """Execute the program."""
+    global goptions
+    goptions = options
+
+    if options.install_directory is not None:
+        # Check that the install directory exists.
+        if not os.path.exists(options.install_directory):
+            sys.exit(
+                f"Error: given install directory '{options.install_directory}' does not exist")
+
+    # Locate the packages to document.
+    packages = find_packages_allowing_duplicates(options.package_path)
+    if len(packages) == 0:
+        logger_scan.error(f'No packages found in subdirectories of {options.package_path}')
+        exit(1)
+    max_packages = int(options.max_packages)
+    if len(packages) > max_packages:
+        packages = packages[0:max_packages]
+
+    packages_total = len(packages)
+    packages_done = 0
+    logger_scan.info(f'Processing {packages_total} packages')
+    batch_packages = []
+    batch_no = 0
+    failed_packages = []
+    while len(packages) > 0:
+        batch_no += 1
+        batch_packages.clear()
+        packages = list(packages.values())
+        for i in range(len(packages)):
+            batch_packages.append(packages.pop())
+            if len(batch_packages) >= BATCH_SIZE:
```

Comment:
At a broader view, what is the value of the batching for running this? It basically forces the mp.Pool to be reset periodically. And with the current default it's not expected to ever be hit. As such might it make sense to reduce the complexity and not batch things before the pool. If someone wants to debug a build/scan they can change the argument to a specific subfolder which is where the problem is being encountered more intuitively than trying to tune the batch size.

Reply:
The batching originally got added during development, when things were dying and it was really hard to figure out where and why. With small batch sizes, the problem could be isolated to a small number of possible packages to look for problems. As a debugging tool, it was easy enough to just change the value in code to enable batching. I don't really see the value except in debugging, hence it was not exposed to the user.

I have not had issues lately with it dying though, so I think I've caught all of the problems with other tools (adding a timeout, running Sphinx as a subprocess again, better logging of individual package outputs.) But I left it in since it was useful during initial debugging. But I don't feel strongly either, as --max-packages and --timeout mostly serve that need. I think I would rather remove it than expose it to the user though, as it really is not of any value except to the rosdoc2 developer. So unless you have another viewpoint, I'll remove the batching.

Reply:
Yeah, I think now that it's working well, it would be better to remove it and keep the much simpler flow of build and list and execute it in a pool.